### PR TITLE
refactor: Remove safer_staticfiles, replace STATICFILES_DIRS with COMPONENTS.dirs, support `[app]/components`

### PR DESCRIPTION
**Pull request by JuroOravec**
As the title says, this one removes `safer_staticfiles`, and instead uses the staticfile's Finders to find additional static files. So people can have their usual static media configured by `STATICFILES_DIRS`, and components configured via `COMPONENTS.dirs`, independently.

`STATICFILES_DIRS` will still work with django_components until v1 though.

Closes https://github.com/EmilStenstrom/django-components/issues/610

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,6 +71,13 @@ And this is what gets rendered (plus the CSS and Javascript you've specified):
 
 ## Release notes
 
+🚨📢 **Version 0.100**
```
Comment (JuroOravec) [inline README.md:74]: Skipped to a round version number as this is a more fundamental change

File: README.md [hunk 2]
```diff
@@ -191,14 +198,20 @@ _You are advised to read this section before using django-components in producti
```
Comment (JuroOravec) [inline README.md:198]: In v1, this section on static files could be removed, but until then it makes sense to keep it

File: README.md [hunk 3]
```diff
@@ -230,11 +243,11 @@ For a step-by-step guide on deploying production server with static files,
 
 ## Installation
 
-1. Install the app into your environment:
+1. Install `django_components` into your environment:
```
Comment (JuroOravec) [inline README.md:246]: Changed the wording here and there, so the purpose of each step is clearer. Also, what was previously step 5 (`STATICFILES_DIRS`) is now split into steps 4 and 6 (`COMPONENTS.dirs` and `STATICFILES_FINDERS`)

File: sampleproject/sampleproject/settings.py
Comment (JuroOravec) [inline sampleproject/sampleproject/settings.py:1]: This file shows nicely the changes required after this MR

File: src/django_components/finders.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+from django.contrib.staticfiles.utils import get_files
+from django.core.checks import CheckMessage, Error, Warning
+from django.core.files.storage import FileSystemStorage
+from django.utils._os import safe_join
+
+from django_components.app_settings import app_settings
+from django_components.template_loader import get_dirs
+
+# To keep track on which directories the finder has searched the static files.
+searched_locations = []
```
Comment (JuroOravec) [inline src/django_components/finders.py:16]: This global variable, and most of this file overall, was based on / taken from FileSystemFinder

File: src/django_components/finders.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    Differences:
+    - This finder uses ``COMPONENTS.dirs`` setting to locate files instead of ``STATICFILES_DIRS``.
+    - This finder ignores Python and HTML files (``.py``, ``.pyc``, ``.html``, ``.django``)
+    - If ``COMPONENTS.dirs`` is not set or empty, defaults to ``settings.BASE_DIR / "components"``
+    """
+
+    IGNORE_PATTERNS = StaticFilesConfig.ignore_patterns + ["*.html", "*.django", "*.py", "*.pyc"]
+
+    def __init__(self, app_names: Any = None, *args: Any, **kwargs: Any) -> None:
+        component_dirs = [str(p) for p in get_dirs()]
```
Comment (JuroOravec) [inline src/django_components/finders.py:38]: It's still the Loader class where we define how we search the filesystem for component files. Here, inside the Finder, we use `get_dirs()`, which calls / uses the loader.

File: src/django_components/template_loader.py [hunk 6]
```diff
@@ -3,59 +3,103 @@
...
+
+        component_dirs = component_dirs or [settings.BASE_DIR / "components"]
 
         logger.debug(
             "Template loader will search for valid template dirs from following options:\n"
             + "\n".join([f" - {str(d)}" for d in component_dirs])
         )
 
-        directories: Set[Path] = set()
+        # Add `[app]/components` to the directories
```
Comment (JuroOravec) [inline src/django_components/template_loader.py:60]: This re-adds support for app-level components dir

File: tests/test_finders.py [hunk 7]
```diff
@@ -60,44 +60,57 @@ def do_collect():
...
+    "INSTALLED_APPS": ("django_components", "django.contrib.staticfiles"),
 }
 
 
-# Check that .py and .html files are INCLUDED with the original staticfiles app
+# Check that the component files are NOT loaded when our finder is NOT added
 @override_settings(
     **common_settings,
-    INSTALLED_APPS=("django_components", "django.contrib.staticfiles"),
+    STATICFILES_FINDERS=[
```
Comment (JuroOravec) [inline tests/test_finders.py:76]: I reused the tests for `safer_staticfiles` to check that we indeed load only JS and CSS

File: tests/test_template_loader.py [hunk 8]
```diff
@@ -24,7 +24,10 @@ def test_get_dirs__base_dir(self):
             sorted(dirs),
             sorted(
                 [
+                    # Top-level /components dir
                     Path(__file__).parent.resolve() / "components",
+                    # App-level /components dir
+                    Path(__file__).parent.resolve() / "test_app" / "components",
```
Comment (JuroOravec) [inline tests/test_template_loader.py:30]: This is where we test the app-level components dirs. For this to work, I added a Django app to `/tests/test_app` dir

File: README.md [hunk 9]
```diff
@@ -71,6 +71,14 @@ And this is what gets rendered (plus the CSS and Javascript you've specified):
 
 ## Release notes
 
+🚨📢 **Version 0.100**
+- BREAKING CHANGE:
+    - `django_components.safer_staticfiles` app was removed. It is no longer needed.
+    - Installation change: Instead of defining component directories in `STATICFILES_DIRS`, set them to `COMPONENTS.dirs`.
+    - Installation change: You now must define `STATICFILES_FINDERS`
+    - [See here how to migrate your settings.py](https://github.com/EmilStenstrom/django-components/pull/652/files#diff-59eb658f2c5573a64306ac0adc96405344e029fd2df51573555c82e44b4d7062)
```
Comment (EmilStenstrom) [inline README.md:79]: I think this is too hard to follow. Don't think everyone is comfortable following git diffs around I'm afraid. Maybe we need a new page to link to?

File: README.md [hunk 10]
```diff
@@ -243,13 +257,32 @@ For a step-by-step guide on deploying production server with static files,
...
+            ...,
+            os.path.join(BASE_DIR, "components"),
+        ],
+   }
+   ```
+
+   If `COMPONENTS.dirs` is omitted or empty, django-components will by default look for a top-level `/components` directory,
+   `{BASE_DIR}/components`.
+
+   Irrespective of `COMPONENTS.dirs`, django_components will also load components from app-level `/components` directories, e.g. `my-app/components/`.
```
Comment (EmilStenstrom) [inline README.md:280]: Hm. So I can set the directory for the global componets, but not for the app level ones? And searching for them can't be disabled?

File: README.md [hunk 11]
```diff
@@ -243,13 +257,32 @@ For a step-by-step guide on deploying production server with static files,
...
+
+   NOTE: The input to `COMPONENTS.dirs` are the same as for `STATICFILES_DIRS`, and the paths must be full paths. [See Django docs](https://docs.djangoproject.com/en/5.0/ref/settings/#staticfiles-dirs).
+
+
+5. Next, to make Django load component HTML files as Django templates, modify `TEMPLATES` section of settings.py as follows:
```
Comment (EmilStenstrom) [inline README.md:285]: Since the defaults are so good, I think we can remove this from the basic setup. Everyone should be good with the defaults.

File: README.md [hunk 12]
```diff
@@ -278,19 +311,18 @@ For a step-by-step guide on deploying production server with static files,
...
-   ```
+6. Lastly, be able to serve the component JS and CSS files as static files, modify `STATICFILES_FINDERS` section of settings.py as follows:
 
-   If `STATICFILES_DIRS` is omitted or empty, django-components will by default look for
-   `{BASE_DIR}/components`
+```py
+STATICFILES_FINDERS = [
+	# Default finders
+    "django.contrib.staticfiles.finders.FileSystemFinder",
+    "django.contrib.staticfiles.finders.AppDirectoriesFinder",
```
Comment (EmilStenstrom) [inline README.md:320]: Indentation.

File: src/django_components/finders.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+# This is what makes it possible to define JS and CSS files in the directories as
+# defined by `COMPONENTS.dirs`, but still use the JS / CSS files with `static()` or
+# `collectstatic` command.
+class ComponentsFileSystemFinder(BaseFinder):
+    """
+    A static files finder based on ``FileSystemFinder``.
+
+    Differences:
+    - This finder uses ``COMPONENTS.dirs`` setting to locate files instead of ``STATICFILES_DIRS``.
+    - This finder ignores Python and HTML files (``.py``, ``.pyc``, ``.html``, ``.django``)
```
Comment (EmilStenstrom) [inline src/django_components/finders.py:31]: I've never seen the .django ending, is that is use somewhere?

File: src/django_components/finders.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    - If ``COMPONENTS.dirs`` is not set or empty, defaults to ``settings.BASE_DIR / "components"``
+    """
+
+    IGNORE_PATTERNS = StaticFilesConfig.ignore_patterns + ["*.html", "*.django", "*.py", "*.pyc"]
```
Comment (EmilStenstrom) [inline src/django_components/finders.py:35]: I'm a little worried about the security implications of this. If someone accidentally would put some other file type in these dirs, they could get exposed on the internet.

At the same time, explicitly including file types is hard, because the full list of potential static files is very large...

Any ideas of how to make this both safe and future-proof?

**Review by EmilStenstrom — COMMENTED** [review]
Exciting! :D

Some docs questions and a bigger point about how to make the searching for files both safe for avoiding mistakes and future-proof.

File: src/django_components/finders.py (hunk 14 above)
Comment (EmilStenstrom) [reply src/django_components/finders.py:35]: Maybe a combo of both?

```python
ALLOWED_EXTENSIONS = getattr(settings, 'COMPONENTS_ALLOWED_EXTENSIONS', 
                             ['.css', '.js', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.woff', '.woff2', '.ttf', '.eot'])
FORBIDDEN_EXTENSIONS = ['.py', '.pyc', '.html', '.django', '.env', '.db', '.sqlite3']
```

File: README.md (hunk 11 above)
Comment (JuroOravec) [reply README.md:285]: We can't, we still need to insert `django_components.template_loader.Loader` so we can let Django handle the loading and caching of the HTML templates.

File: src/django_components/finders.py (hunk 14 above)
Comment (JuroOravec) [reply src/django_components/finders.py:35]: I didn't bother to change this, but agree with you, explicitly listing allowed extensions is better than the opposite. The reason why I used only ignored patterns was because then we can delegate the filtering to [`django.contrib.staticfiles.utils.get_files`](https://github.com/django/django/blob/main/django/contrib/staticfiles/utils.py#L16), which acepts a list of glob patterns.

But globbing doesn't support negations, so one cannot search for "everything except `.html`". So we can't use `get_files` to get only files with allowed patterns. And so we need to do the filtering ourselves.

And TBH, if we're doing the filtering ourselves, I would go with regex over globs, so users can be more expressive, e.g. maybe they would want to exclude CSS only from certain components, or omit static files that start with underscore.

But I just tried it, and writing regex for patterns like "everything except `.html`, `.py`, or `.pyc`" is not very easy to read. So I would still go with as you suggested, specifying both allowed and forbidden patterns. One approach I've seen that I liked was that first all files from allowed patterns are found, and then any forbidden files are removed. So, in other words, if a file was matched successfully against both `ALLOWED_EXTENSIONS` and `FORBIDDEN_EXTENSIONS`, then the file will be OMITTED.

```py
COMPONENTS = {
	"allowed_static_files": [
		r"\.{css|js|png|jpg|jpeg|svg}$",
	],
	"forbidden_static_files": [
		r"\.{html|py|pyc}$",
	],
}
```

The defaults would then be:
```py
COMPONENTS = {
	"allowed_static_files": [
		# Match all files
		".*",
	],
	"forbidden_static_files": [
		# Exclude .html, .py, .pyc
		r"\.{html|py|pyc}$",
	],
}
```

File: src/django_components/finders.py (hunk 13 above)
Comment (JuroOravec) [reply src/django_components/finders.py:31]: I've seen it used in this VSCode extension https://marketplace.visualstudio.com/items?itemName=junstyle.vscode-django-support.

They use `.django`, `.dj`, `.tpl`, but I forgot about the latter two.

File: src/django_components/app_settings.py [hunk 15]
```diff
@@ -118,6 +126,25 @@ def RELOAD_ON_TEMPLATE_CHANGE(self) -> bool:
     def TEMPLATE_CACHE_SIZE(self) -> int:
         return self.settings.get("template_cache_size", 128)
 
+    @property
+    def STATIC_FILES_ALLOWED(self) -> List[str]:
+        default_static_files = [
+            # Allow suffixes: .js, .css
+            r"\.(?:js|css)$",
```
Comment (EmilStenstrom) [inline src/django_components/app_settings.py:133]: Should we not also allow images by default?

**Review by EmilStenstrom — COMMENTED** [review]
This is great work, as always! :)

1) I personally would have gone with just a list of file endings instead of regex. Developers are not as comfortable with regex as experienced developers think. They might mess up that ".html" means "[anything]html", they might mess up the raw string, the parenthesis, not understand how they would allow wildcard to the list? [When I advocated that Django move away from regex urlpatterns](https://groups.google.com/g/django-developers/c/u6sQax3sjO4/m/l3nvZCuuAAAJ) to path urlpatterns that was the exact same argument.

2) A generally question: What's a good default, shouldn't we allow images? Fonts?

File: src/django_components/app_settings.py (hunk 15 above)
Comment (EmilStenstrom) [reply src/django_components/app_settings.py:133]: Suggestion:

```python
default_static_files = [".css", ".gif", ".ico", ".jpeg", ".jpg", ".js", ".png", ".svg", ".webp"]
````

Comment (JuroOravec) [reply src/django_components/app_settings.py:133]: Good question. We can add [common image types as defined in MDN](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#common_image_file_types).

As for fonts, at first my reaction was that these should be probably at the root of the project, outside of components. But one could have an "app" component that would hold fonts and such, so I can see that now. We can go with the ones listed here https://stackoverflow.com/q/30572159/9788634

Comment (JuroOravec) [reply src/django_components/app_settings.py:133]: For the regex patterns, what do you say if we supported both? That preferred way would be to pass in strings, which would be evaluated as suffixes like `.html`. But one could also pass an instance of `re.Pattern` (AKA the result of `re.compile`) for more control.

Comment (JuroOravec) [reply src/django_components/app_settings.py]: Updated, see the changes here https://github.com/EmilStenstrom/django-components/pull/652/files/766fd047711463a32f2910cf578dbfcb27b5cd9d..16b418d8d78e56b78a68f0a6f184ade8dc6daec4

**Review by EmilStenstrom — APPROVED** [review]
Looks great! <3
